chia/data_layer/download_data.py

```diff
@@ -0,0 +1,184 @@
+import asyncio
+import logging
+import os
+import time
+from pathlib import Path
+from typing import List, Optional
+
+import aiohttp
+from typing_extensions import Literal
+
+from chia.data_layer.data_layer_util import NodeType, Root, SerializedNode, ServerInfo, Status
+from chia.data_layer.data_store import DataStore
+from chia.types.blockchain_format.sized_bytes import bytes32
+
+
+def get_full_tree_filename(tree_id: bytes32, node_hash: bytes32, generation: int) -> str:
+    return f"{tree_id}-{node_hash}-full-{generation}-v1.0.dat"
+
+
+def get_delta_filename(tree_id: bytes32, node_hash: bytes32, generation: int) -> str:
+    return f"{tree_id}-{node_hash}-delta-{generation}-v1.0.dat"
+
+
+def is_filename_valid(filename: str) -> bool:
+    split = filename.split("-")
+
+    try:
+        raw_tree_id, raw_node_hash, file_type, raw_generation, raw_version, *rest = split
+        tree_id = bytes32(bytes.fromhex(raw_tree_id))
+        node_hash = bytes32(bytes.fromhex(raw_node_hash))
+        generation = int(raw_generation)
+    except ValueError:
+        return False
+
+    if len(rest) > 0:
+        return False
+
+    # TODO: versions should probably be centrally defined
+    if raw_version != "v1.0.dat":
+        return False
+
+    if file_type not in {"delta", "full"}:
+        return False
+
+    generate_file_func = get_delta_filename if file_type == "delta" else get_full_tree_filename
+    reformatted = generate_file_func(tree_id=tree_id, node_hash=node_hash, generation=generation)
+
+    return reformatted == filename
+
+
+async def insert_into_data_store_from_file(
+    data_store: DataStore,
+    tree_id: bytes32,
+    root_hash: Optional[bytes32],
+    filename: Path,
+) -> None:
+    with open(filename, "rb") as reader:
+        while True:
+            chunk = b""
+            while len(chunk) < 4:
```

Comment:
If we switch to `struct.unpack`, we could version the format strings to match our data store version.

Reply:
use `struct.unpack` only for this line:`size = int.from_bytes(chunk, byteorder="big")`?